eth_utils/abi.py

```diff
@@ -1,68 +1,793 @@
+    ]
+
+
+def filter_abi_by_type(type: str, contract_abi: ABI) -> List[ABIElement]:
+    """
+    Return a list of each ``ABIElement`` that is of type ``_type``.
+
+    :param type: Type of ABI element to filter by.
+    :type type: `str`
+    :param contract_abi: Contract ABI.
+    :type contract_abi: `ABI`
+    :return: List of ABI elements of the specified type.
+    :rtype: `List[ABIElement]`
+
+    .. doctest::
+
+        >>> from eth_utils import filter_abi_by_type
+        >>> abi = [
+        ...   {"type": "function", "name": "myFunction", "inputs": [], "outputs": []},
+        ...   {"type": "function", "name": "myFunction2", "inputs": [], "outputs": []},
+        ...   {"type": "event", "name": "MyEvent", "inputs": []}
+        ... ]
+        >>> filter_abi_by_type("function", abi)
+        [{'type': 'function', 'name': 'myFunction', 'inputs': [], 'outputs': []}, \
+{'type': 'function', 'name': 'myFunction2', 'inputs': [], 'outputs': []}]
+    """
+    return [abi for abi in contract_abi if abi["type"] == type]
+
+
+def get_all_function_abis(contract_abi: ABI) -> Sequence[ABIFunction]:
+    """
+    Return interfaces for each function in the contract ABI.
+
+    :param abi: Contract ABI.
+    :type abi: `ABI`
+    :return: List of ABIs for each function interface.
+    :rtype: `list[ABIFunction]`
+
+    .. doctest::
+
+        >>> from eth_utils import get_all_function_abis
+        >>> contract_abi = [
+        ...   {"type": "function", "name": "myFunction", "inputs": [], "outputs": []},
+        ...   {"type": "function", "name": "myFunction2", "inputs": [], "outputs": []},
+        ...   {"type": "event", "name": "MyEvent", "inputs": []}
+        ... ]
+        >>> get_all_function_abis(contract_abi)
+        [{'type': 'function', 'name': 'myFunction', 'inputs': [], 'outputs': []}, \
+{'type': 'function', 'name': 'myFunction2', 'inputs': [], 'outputs': []}]
+    """
+    return [
+        cast(ABIFunction, function_abi)
+        for function_abi in filter_abi_by_type("function", contract_abi)
+    ]
+
+
+def get_all_event_abis(contract_abi: ABI) -> Sequence[ABIEvent]:
+    """
+    Return interfaces for each event in the contract ABI.
+
+    :param abi: Contract ABI.
+    :type abi: `ABI`
+    :return: List of ABIs for each event interface.
+    :rtype: `list[ABIEvent]`
+
+    .. doctest::
+
+        >>> from eth_utils import get_all_event_abis
+        >>> contract_abi = [
+        ...   {"type": "function", "name": "myFunction", "inputs": [], "outputs": []},
+        ...   {"type": "function", "name": "myFunction2", "inputs": [], "outputs": []},
+        ...   {"type": "event", "name": "MyEvent", "inputs": []}
+        ... ]
+        >>> get_all_event_abis(contract_abi)
+        [{'type': 'event', 'name': 'MyEvent', 'inputs': []}]
+    """
+    return [
+        cast(ABIEvent, event) for event in filter_abi_by_type("event", contract_abi)
+    ]
+
+
+def get_normalized_abi_inputs(
+    function_abi: Union[ABIFunction, ABIError],
```

Comment:
Are these types of methods purposefully narrowing the scope to just functions, and skipping `constructor`, `fallback`, and `receive` - i.e. `ABICallable`?

Reply:
Should add `constructor` to these, that's a good call. There are no inputs for `fallback` and `receive` types, but if this were `ABIElement` instead of `Union[ABIFunction, ABIError]` there could be logic to throw a `ValueError` if an invalid type is passed in. At least that's what I've done for other methods like `get_abi_input_names`.

Let me know if you think we should change this to raise an exception and allow all `ABIElement` types. If not, I'll just change the argument type to `Union[ABIFunction, ABIConstructor, ABIError]`.

Reply:
I think I'd lean towards `ABIElement` and throw an error, especially if that's a pattern already in use elsewhere.

Reply:
Updated a few places where this was not consistent. See https://github.com/ethereum/eth-utils/pull/271/commits/0c49bc6c37f59f1252a07de70f52f2f5eaf5a859